File: aihound/scanners/amazon_q.py
"""Scanner for Amazon Q Developer / AWS credentials."""

from __future__ import annotations

import configparser
import json
from pathlib import Path

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType,
)
from aihound.core.platform import detect_platform, Platform, get_home, get_wsl_windows_home
from aihound.core.redactor import mask_value
from aihound.core.permissions import get_file_permissions, get_file_owner, assess_risk, get_file_mtime, describe_staleness
from aihound.remediation import hint_manual, hint_rotate_credential
from aihound.scanners import register


@register
class AmazonQScanner(BaseScanner):
    def name(self) -> str:
        return "Amazon Q / AWS"

    def slug(self) -> str:
        return "amazon-q"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        for path in self._get_credential_paths(plat):
            self._scan_aws_credentials(path, result, show_secrets)

        for path in self._get_sso_cache_paths(plat):
            self._scan_sso_cache(path, result, show_secrets)

        return result

    def _get_credential_paths(self, plat: Platform) -> list[Path]:
        paths = []
        home = get_home()

        paths.append(home / ".aws" / "credentials")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".aws" / "credentials")

        return paths

    def _get_sso_cache_paths(self, plat: Platform) -> list[Path]:
        paths = []
        home = get_home()

        paths.append(home / ".aws" / "sso" / "cache")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".aws" / "sso" / "cache")

        return paths

    def _scan_aws_credentials(
        self, path: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not path.exists():
            return

        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)
        storage = StorageType.PLAINTEXT_INI

        config = configparser.ConfigParser()
        try:
            config.read(str(path))
        except (configparser.Error, OSError) as e:
            result.errors.append(f"Failed to parse {path}: {e}")
            return

        for section in config.sections():
            for key in ("aws_access_key_id", "aws_secret_access_key", "aws_session_token"):
                value = config.get(section, key, fallback=None)
                if value:
                    notes = [f"AWS profile: [{section}]"]
                    if mtime:
                        notes.append(f"File last modified: {describe_staleness(mtime)}")
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=key,
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(value, show_full=show_secrets),
                        raw_value=value if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation="Use AWS SSO or IAM roles instead of long-lived access keys",
                        remediation_hint=hint_manual(
                            "Use AWS SSO or IAM roles instead of long-lived access keys",
                            suggested_commands=["aws configure sso"],
                        ),
                        notes=notes,
                    ))

    def _scan_sso_cache(
        self, cache_dir: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not cache_dir.exists():
            return

        for json_file in cache_dir.glob("*.json"):
            perms = get_file_permissions(json_file)
            owner = get_file_owner(json_file)
            mtime = get_file_mtime(json_file)

            try:
                data = json.loads(json_file.read_text(encoding="utf-8"))
            except (json.JSONDecodeError, OSError):
                continue

            if isinstance(data, dict):
                access_token = data.get("accessToken")
                if access_token and isinstance(access_token, str):
                    notes = ["AWS SSO cached token"]
                    if mtime:
                        notes.append(f"File last modified: {describe_staleness(mtime)}")
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type="sso_access_token",
                        storage_type=StorageType.PLAINTEXT_JSON,
                        location=str(json_file),
                        exists=True,
                        risk_level=assess_risk(StorageType.PLAINTEXT_JSON, json_file),
                        value_preview=mask_value(access_token, show_full=show_secrets),
                        raw_value=access_token if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation="Rotate SSO tokens regularly",
                        remediation_hint=hint_rotate_credential("aws-sso", "Rotate SSO tokens regularly"),
                        notes=notes,
                    ))

File: Other Versions/Go/scanners/amazon_q.go
package scanners

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

type amazonQScanner struct{}

func init() {
	Register(&amazonQScanner{})
}

func (s *amazonQScanner) Name() string      { return "Amazon Q / AWS" }
func (s *amazonQScanner) Slug() string       { return "amazon-q" }
func (s *amazonQScanner) IsApplicable() bool { return true }

func (s *amazonQScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, path := range s.getCredentialPaths(plat) {
		s.scanAWSCredentials(path, &result, showSecrets)
	}

	for _, path := range s.getSSOCachePaths(plat) {
		s.scanSSOCache(path, &result, showSecrets)
	}

	return result
}

func (s *amazonQScanner) getCredentialPaths(plat core.Platform) []string {
	home := core.GetHome()
	paths := []string{filepath.Join(home, ".aws", "credentials")}

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".aws", "credentials"))
		}
	}

	return paths
}

func (s *amazonQScanner) getSSOCachePaths(plat core.Platform) []string {
	home := core.GetHome()
	paths := []string{filepath.Join(home, ".aws", "sso", "cache")}

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".aws", "sso", "cache"))
		}
	}

	return paths
}

// scanAWSCredentials parses an INI-format AWS credentials file manually.
func (s *amazonQScanner) scanAWSCredentials(path string, result *core.ScanResult, showSecrets bool) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	storage := core.PlaintextINI

	// Simple manual INI parsing
	sections := parseINI(string(raw))

	awsKeys := []string{"aws_access_key_id", "aws_secret_access_key", "aws_session_token"}

	for sectionName, kvs := range sections {
		for _, key := range awsKeys {
			value, ok := kvs[key]
			if !ok || value == "" {
				continue
			}

			rawValue := ""
			if showSecrets {
				rawValue = value
			}
			notes := []string{fmt.Sprintf("AWS profile: [%s]", sectionName)}
			if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
				notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
			}
			result.Findings = append(result.Findings, core.CredentialFinding{
				ToolName:        s.Name(),
				CredentialType:  key,
				StorageType:     storage,
				Location:        path,
				Exists:          true,
				RiskLevel:       core.AssessRisk(storage, path),
				ValuePreview:    core.MaskValue(value, showSecrets),
				RawValue:        rawValue,
				FilePermissions: perms,
				FileOwner:       owner,
				FileModified:    core.GetFileMtime(path),
				Remediation:     "Use AWS SSO or IAM roles instead of long-lived access keys",
				RemediationHint: remediation.HintManual(
					"Use AWS SSO or IAM roles instead of long-lived access keys",
					map[string]any{"suggested_commands": []string{"aws configure sso"}},
				),
				Notes:           notes,
			})
		}
	}
}

// parseINI does minimal INI parsing: [section] headers and key=value pairs.
func parseINI(content string) map[string]map[string]string {
	sections := make(map[string]map[string]string)
	currentSection := ""

	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") || strings.HasPrefix(line, ";") {
			continue
		}

		// Section header
		if strings.HasPrefix(line, "[") && strings.HasSuffix(line, "]") {
			currentSection = strings.TrimSpace(line[1 : len(line)-1])
			if _, ok := sections[currentSection]; !ok {
				sections[currentSection] = make(map[string]string)
			}
			continue
		}

		// Key=value pair
		if idx := strings.Index(line, "="); idx >= 0 && currentSection != "" {
			key := strings.TrimSpace(line[:idx])
			value := strings.TrimSpace(line[idx+1:])
			sections[currentSection][key] = value
		}
	}

	return sections
}

func (s *amazonQScanner) scanSSOCache(cacheDir string, result *core.ScanResult, showSecrets bool) {
	if _, err := os.Stat(cacheDir); err != nil {
		return
	}

	matches, _ := filepath.Glob(filepath.Join(cacheDir, "*.json"))
	for _, jsonFile := range matches {
		data, err := os.ReadFile(jsonFile)
		if err != nil {
			continue
		}

		perms := core.GetFilePermissions(jsonFile)
		owner := core.GetFileOwner(jsonFile)

		var obj map[string]interface{}
		if err := json.Unmarshal(data, &obj); err != nil {
			continue
		}

		accessToken, ok := obj["accessToken"].(string)
		if !ok || accessToken == "" {
			continue
		}

		rawValue := ""
		if showSecrets {
			rawValue = accessToken
		}
		notes := []string{"AWS SSO cached token"}
		if mtime := core.GetFileMtimeTime(jsonFile); !mtime.IsZero() {
			notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
		}
		result.Findings = append(result.Findings, core.CredentialFinding{
			ToolName:        s.Name(),
			CredentialType:  "sso_access_token",
			StorageType:     core.PlaintextJSON,
			Location:        jsonFile,
			Exists:          true,
			RiskLevel:       core.AssessRisk(core.PlaintextJSON, jsonFile),
			ValuePreview:    core.MaskValue(accessToken, showSecrets),
			RawValue:        rawValue,
			FilePermissions: perms,
			FileOwner:       owner,
			FileModified:    core.GetFileMtime(jsonFile),
			Remediation:     "Rotate SSO tokens regularly",
			RemediationHint: remediation.HintRotateCredential("aws-sso", "Rotate SSO tokens regularly"),
			Notes:           notes,
		})
	}
}
